### DISCUSSION TITLE
Clock reset to 00:00

### DISCUSSION THREAD

**Discussion by retomanic** [Makuna/Rtc#139] [Q&A] [answered]
Hi, I've been trying to figure out this case with my RTC DS1302 module.

When I boot the Arduino Uno up, serial monitor would begin to show the accurate clock to my current time, and then it begins to reset back to 00:00 in a loop.

Serial monitor screen:
```
compiled: May  4 202223:36:13
05/04/2022 23:36:13
RTC lost confidence in the DateTime!
RTC is newer than compile time. (this is expected)
00/00/2000 00:00:00RTC lost confidence in the DateTime!
00/00/2000 00:00:00RTC lost confidence in the DateTime!
00/00/2000 00:00:00RTC lost confidence in the DateTime!
00/00/2000 00:00:00RTC lost confidence in the DateTime!
00/00/2000 00:00:00RTC lost confidence in the DateTime!
```

This are the codes that I am using:
```
`#include <ThreeWire.h>  
#include <RtcDS1302.h>
ThreeWire myWire(6,7,8); // IO, SCLK, CE
RtcDS1302<ThreeWire> Rtc(myWire);
void setup () {
Serial.begin(115200);
    Serial.print("compiled: ");
    Serial.print(__DATE__);
    Serial.println(__TIME__);
    Rtc.Begin();
    RtcDateTime compiled = RtcDateTime(__DATE__, __TIME__);
    printDateTime(compiled);
    Serial.println();
    if (!Rtc.IsDateTimeValid()) {
        // Common Causes: 1) first time you ran and the device wasn't running yet 2) the battery on the device is low or even missing
        Serial.println("RTC lost confidence in the DateTime!");
        Rtc.SetDateTime(compiled);
    }
    if (Rtc.GetIsWriteProtected()) {
        Serial.println("RTC was write protected, enabling writing now");
        Rtc.SetIsWriteProtected(false);
    }
    if (!Rtc.GetIsRunning()) {
        Serial.println("RTC was not actively running, starting now");
    Rtc.SetIsRunning(true);
    }
    RtcDateTime now = Rtc.GetDateTime();
    if (now < compiled) {
        Serial.println("RTC is older than compile time!  (Updating DateTime)");
        Rtc.SetDateTime(compiled);
 }
    else if (now > compiled)
        Serial.println("RTC is newer than compile time. (this is expected)");
    else
        Serial.println("RTC is the same as compile time! (not expected but all is fine)");
}
void loop () {
    RtcDateTime now = Rtc.GetDateTime();
    printDateTime(now);
    if (!now.IsValid()) {
        // Common Causes: 1) the battery on the device is low or even missing and the power line was disconnected
        Serial.println("RTC lost confidence in the DateTime!");
    }
    delay(1000);
} 
#define countof(a) (sizeof(a) / sizeof(a[0]))
void printDateTime(const RtcDateTime& dt) {
    char datestring[20];
    snprintf_P(datestring, 
            countof(datestring),
            PSTR("%02u/%02u/%04u %02u:%02u:%02u"),
            dt.Month(), dt.Day(), dt.Year(),
            dt.Hour(), dt.Minute(), dt.Second() );
    Serial.print(datestring);
}
```

Please help..... I've changed the batteries to a new one, I've toggled the wires and I've did all I could. I'm worried it's a setup I've missed etc.

**Comment by Makuna** [comment]
Your debug output doesn't match your statement about showing correct time and then clearing to zero.  

at the last line in setup, add 
```
printDateTime(now);
Serial.println(" Setup done");
```

**Reply by retomanic** [reply]
Thank you for your prompt response.
Sorry, instead of zero it prints out this: "00/00/2000 00:00:00RTC lost confidence in the DateTime!". It shows correct time here in this line: "compiled: May  4 202223:36:13
05/04/2022 23:36:13
RTC lost confidence in the DateTime!"

I've tried adding :
"printDateTime(now);
Serial.println(" Setup done");"
 in the last line of Void setup {}

here:     else if (now > compiled)
        Serial.println("RTC is newer than compile time. (this is expected)");
    else
        Serial.println("RTC is the same as compile time! (not expected but all is fine)");
           printDateTime(now);
Serial.println(" Setup done");
}

But still to no avail.

**Reply by Makuna** [reply]
And after you added the lines I asked for, what is the output?  Copy it here.

The lines you state above in the debug output are JUST the compiled date time after being converted into an RtcDateTime, not the values from the actual chip.
```
    Serial.print("compiled: ");
    Serial.print(__DATE__);
    Serial.println(__TIME__);
    Rtc.Begin();
    RtcDateTime compiled = RtcDateTime(__DATE__, __TIME__);
    printDateTime(compiled); // <<<<---- this is the compiled date time after converting, not the value from the chip
    Serial.println();
```

**Reply by retomanic** [reply]
So I've added the lines into this script:
`

#include <LiquidCrystal_I2C.h>
#include <ThreeWire.h>  
#include <RtcDS1302.h>
ThreeWire myWire(6,7,8); // IO, SCLK, CE
RtcDS1302<ThreeWire> Rtc(myWire);
LiquidCrystal_I2C lcd(0x27,16,2);  // set the LCD address to 0x27 for a 16 chars and 2 line display
void setup () {

  
Serial.begin(9600);
    Serial.print("compiled: ");
    Serial.print(__DATE__);
    Serial.println(__TIME__);
    Rtc.Begin();
    RtcDateTime compiled = RtcDateTime(__DATE__, __TIME__);
    printDateTime(compiled);
    Serial.println();
    if (!Rtc.IsDateTimeValid()) {
        // Common Causes: 1) first time you ran and the device wasn't running yet 2) the battery on the device is low or even missing
        Serial.println("RTC lost confidence in the DateTime!");
        Rtc.SetDateTime(compiled);
    }
    if (Rtc.GetIsWriteProtected()) {
        Serial.println("RTC was write protected, enabling writing now");
        Rtc.SetIsWriteProtected(false);
    }
    if (!Rtc.GetIsRunning()) {
        Serial.println("RTC was not actively running, starting now");
    Rtc.SetIsRunning(true);
    }
    RtcDateTime now = Rtc.GetDateTime();
    if (now < compiled) {
        Serial.println("RTC is older than compile time!  (Updating DateTime)");
        Rtc.SetDateTime(compiled);
 }
    else if (now > compiled)
        Serial.println("RTC is newer than compile time. (this is expected)");
    else
        Serial.println("RTC is the same as compile time! (not expected but all is fine)");
           printDateTime(now);
Serial.println(" Setup done");
}
void loop () {
    RtcDateTime now = Rtc.GetDateTime();
    printDateTime(now);
    if (!now.IsValid()) {
        // Common Causes: 1) the battery on the device is low or even missing and the power line was disconnected
        Serial.println("RTC lost confidence in the DateTime!");
    }
    delay(1000);
} 
#define countof(a) (sizeof(a) / sizeof(a[0]))
void printDateTime(const RtcDateTime& dt) {
    char datestring[20];
    snprintf_P(datestring, 
            countof(datestring),
            PSTR("%02u/%02u/%04u %02u:%02u:%02u"),
            dt.Month(), dt.Day(), dt.Year(),
            dt.Hour(), dt.Minute(), dt.Second() );
    Serial.print(datestring);
 
        lcd.init();
        lcd.clear(); 
        lcd.backlight();      // Make sure backlight is on// Print a message on both lines of the LCD.
        lcd.setCursor(2,0);   //Set cursor to character 2 on line 0
        lcd.print("Time is now");
        lcd.setCursor(2,1);   //Set cursor to character 2 on line 0
        lcd.print(datestring);
}`

And the output would be this:
![image](https://user-images.githubusercontent.com/91540892/166891517-caa6b332-69cd-438a-99d7-ac079fdb65f4.png)

**Reply by Makuna** [reply]
So, do you see that the clock is not resetting, it just never has the correct time? 
There is very little error handling that can be done with the communications if the wiring is not correct; this is why it is the usual suspect.

**Reply by retomanic** [reply]
Thanks for the prompt feedback. Sure, allow me to tweak the wiring now. Will update if it works.

**Reply by retomanic** [reply, marked as answer]
It seems like it works now after changing Clk, Dat, and Rst to 7, 6, 8 respectively in that order.

Just now that the clock seems to be in a different timezone or something. Would I be able to set timezone to GMT+8?

**Reply by Makuna** [reply]
RtcDateTime doesn't consider timezone at all, and neither do these RTC chips.  As the sketch author, you need to decide if you store local or GMT time.  GMT time is the best approach, then all you have to do is convert at display time to user or user input back and forth.  But this conversion is left to the author.

**Reply by retomanic** [reply]
I understand. It seems to work well now that I have reuploaded the code into the microcontroller.
Makuna, thank you for your ever-kind and prompt feedback. I'm looking to learn if you have created any other libraries for Arduino, had a pleasant time learning from you.
